---
title: "Learn To Code: Problem Solving"
format: revealjs
---

## Learn to code

You can gather information about code through books, docs, blogposts, ...
but the best way to practice is through **problem solving**.

## Resources

There are too many possible examples to show you but I suggest that you find something that motivates you beyond just 'learning to code'

Find something where the subject matter matches your goals.

## What to look for

- approachable
- language agnostic
- interesting
- popular


## What I'd like to show you today

- Advent of Code
- Hanukkah of Data
- Rosalind
- Kaggle


<!-- ## Comp sci / algorithms -->

<!-- - [**Leetcode**](https://leetcode.com/) is one of the most popular -->
<!-- - [**HackerRank**]() is another -->
<!-- - [**codewars**]() works too -->

<!-- ## A taste -->

<!-- If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23. -->

<!-- Find the sum of all the multiples of 3 or 5 below 1000. -->

<!-- ## A taste -->

<!-- If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23. -->

<!-- Find the sum of all the multiples of 3 or 5 below 1000. -->

<!-- *This is relatively easy* -->

## Databases / Querying: Hannukah of Data

[**Hannukah of Data**](https://hanukkah.bluebird.sh/5783/) is a new annual set of problems relating to querying databases and doing fairly straightforward data manipulation.


## Databases / Querying: Hannukah of Data

**Setup**   

- ⬇️ download data   
- 👨🏽‍💻 create an in-memory database   
- 👍 write queries to select data   
- 🤏 manipulate data and filter rows   
- ✅  find a single customer to solve each problem  

## Example


## Bioinformatics: Rosalind

[**Rosalind**](https://rosalind.info)

Solve problems relating to molecular biology: DNA and protein sequencing, sequence alignment, genome assembly, phylogeny, algorithms for classification and clustering,  etc. 

## Simple example

https://rosalind.info/problems/dna/

Given: A DNA string s
 of length at most 1000 nt.
 
 Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in s
.


## Imperative solution

```{r}
#| echo: true
dna <- 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC' |> 
  strsplit('') |>
  unlist()

counts <- c(A = 0, C = 0, G = 0, T = 0)
for (base in dna) {
  counts[base] <- counts[base] + 1
}

print(counts)
```


## Functional solution 

```{r}
#| echo: true
library(tidyverse)

dna <- 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'

lst('A','C','G','T') |> 
  set_names() |>
  map_dbl(~str_count(dna, pattern = .))
```
